src/Metadata/Cip25/V2.purs

```diff
@@ -0,0 +1,348 @@
+    [ "mediaType" /\ toData mediaType ]
+  mbDescription = fold $ entry.description <#> \description ->
+    [ "description" /\ toData description ]
+
+metadataEntryFromData
+  :: MintingPolicyHash
+  -> Cip25TokenName
+  -> PlutusData
+  -> Maybe Cip25MetadataEntry
+metadataEntryFromData policyId assetName contents = do
+  name <- lookupKey "name" contents >>= fromData
+  image <- lookupKey "image" contents >>= fromDataString
+  mediaType <- for (lookupKey "mediaType" contents) fromData
+  description <- for (lookupKey "description" contents) fromDataString
+  files <- for (lookupKey "files" contents) fromData <#> fromMaybe []
+  pure $
+    wrap { policyId, assetName, name, image, mediaType, description, files }
+
+metadataEntryDecodeAeson
+  :: MintingPolicyHash
+  -> Cip25TokenName
+  -> Aeson
+  -> Either JsonDecodeError Cip25MetadataEntry
+metadataEntryDecodeAeson policyId assetName =
+  caseAesonObject errExpectedObject \obj -> do
+    name <- obj .: "name"
+    image <- obj .: "image"
+    mediaType <- obj .:? "mediaType"
+    description <- obj .:? "description" <#> fromMaybe mempty
+    files <- obj .:? "files" >>= \mbFiles ->
+      fromMaybe [] <$> for mbFiles \files ->
+        traverse decodeAeson =<< note (TypeMismatch "files")
+          (Aeson.toArray files)
+    pure $
+      wrap { policyId, assetName, name, image, mediaType, description, files }
+
+newtype Cip25Metadata = Cip25Metadata (Array Cip25MetadataEntry)
+
+derive instance Generic Cip25Metadata _
+derive instance Newtype Cip25Metadata _
+derive instance Eq Cip25Metadata
+
+instance Show Cip25Metadata where
+  show = genericShow
+
+instance MetadataType Cip25Metadata where
+  metadataLabel _ = wrap (BigInt.fromInt 721)
+
+groupEntries
+  :: Array Cip25MetadataEntry -> Array (NonEmptyArray Cip25MetadataEntry)
+groupEntries =
+  groupBy \(Cip25MetadataEntry a) (Cip25MetadataEntry b) ->
+    a.policyId == b.policyId
+
+instance ToMetadata Cip25Metadata where
+  toMetadata (Cip25Metadata entries) = toMetadata $
+    let
+      dataEntries =
+        groupEntries entries <#>
+          \group ->
+            (toMetadata <<< _.policyId <<< unwrap $ NonEmpty.head group) /\
+              (toMetadata <<< toArray <<< flip map group) \entry ->
+                (unwrap entry).assetName /\ metadataEntryToMetadata entry
+      versionEntry = [ toMetadata "version" /\ toMetadata Cip25V2 ]
+    in
+      dataEntries <> versionEntry
+
+instance FromMetadata Cip25Metadata where
+  fromMetadata (MetadataMap mp1) = do
+    entries <- map concat
+      $ map catMaybes
+      $ for (Map.toUnfoldable mp1)
+          \(key /\ assets) ->
+            if key == toMetadata "version" then
+              -- top-level version tag
+              ( if assets == toMetadata Cip25V2 then pure Nothing
+                else Nothing -- fail if version does not match
+              )
+            else Just -- key is policyId
+
+              case assets of
```

Comment:
I think the comment made the formatting odd here, is it possible to move it elsewhere?